modules/core/src/main/java/org/apache/ignite/internal/jdbc/thin/JdbcThinConnection.java

```diff
@@ -1698,6 +1731,49 @@ private void updateAffinityCache(JdbcQueryExecuteRequest qryReq, JdbcResponse re
         }
     }
 
+    /**
+     * Calculates query retries count for given {@param req}.
+     *
+     * @param stickyIo sticky connection, if any.
+     * @param req Jdbc request.
+     *
+     * @return retries count.
+     */
+    private int calculateRetryAttemptsCount(JdbcThinTcpIo stickyIo, JdbcRequest req) {
+        if (! affinityAwareness)
+            return NO_RETRIES;
+
+        if (stickyIo != null)
+            return NO_RETRIES;
+
+        if (req.type() == JdbcRequest.META_TABLES ||
+            req.type() == JdbcRequest.META_COLUMNS ||
+            req.type() == JdbcRequest.META_INDEXES ||
+            req.type() == JdbcRequest.META_PARAMS ||
+            req.type() == JdbcRequest.META_PRIMARY_KEYS ||
+            req.type() == JdbcRequest.META_SCHEMAS ||
+            req.type() == JdbcRequest.CACHE_PARTITIONS)
+            return DFLT_RETRIES_CNT;
+
+        if (req.type() == JdbcRequest.QRY_EXEC) {
+            JdbcQueryExecuteRequest qryExecReq = (JdbcQueryExecuteRequest)req;
+
+            String trimmedQry = qryExecReq.sqlQuery().trim();
+
+            char[] qryCArr = trimmedQry.toCharArray();
```

Comment:
There is no need to have a copy char[]. String.indexOf() can be used instead.
BTW, it is possible to get false positive result if ';' is inside some string const.

Reply:
>There is no need to have a copy char[]. String.indexOf() can be used instead.
Yep, fixed.
>BTW, it is possible to get false positive result if ';' is inside some string const.
That's true. We either should have fully implemented parser on client side or use very simple rules with such false-positive effects. During the discussion, it was decided to start with the simpler version.